### ISSUE TITLE
False positives in modernize-use-default-member-init in clang-tidy 21 when constant is in cpp file and class in a header

### ISSUE THREAD

**Issue by arvid-norlander** [llvm/llvm-project#156412]
Another false positive (but distinct from #156295):

The release notes for [clang-tidy 21](https://releases.llvm.org/21.1.0/tools/clang/tools/extra/docs/ReleaseNotes.html#improvements-to-clang-tidy) reads:
> Improved [modernize-use-default-member-init](https://releases.llvm.org/21.1.0/tools/clang/tools/extra/docs/clang-tidy/checks/modernize/use-default-member-init.html) check by matching arithmetic operations, constexpr and static values, and detecting explicit casting of built-in types within member list initialization.

If you have:

test.h:
```c++
#include <cstdint>

class Example1
{
   explicit Example1(int32_t param);

   int32_t member;
};
```

test.cpp:
```c++
#include "test.h"

namespace {
constexpr double CONSTANT = 2.0;
}

Example1::Example1()
: member(CONSTANT)
{
}

```

clang-tidy-21 will incorrectly report:

```
/home/tmgarn/workspace/reproducer/test.h:7:12: error: use default member initializer for 'member' [modernize-use-default-member-init,-warnings-as-errors]
    7 |    int32_t member;
      |            ^     
      |                  {CONSTANT}
1 warning treated as error
```

If you use `-fix`, it will incorrectly apply this change, but since the constant is defined in the cpp file, the code will not build.

**Comment by arvid-norlander** [comment]
cc @RiverDave (since you were CCed on the other related issue, and absolutely nothing is happening over here).
